### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by AmrSaleh2234**
## 🎯 Summary

-
## 🧾 Related Ticket

- Jira: [ABC-123](https://jira.your-company.com/browse/ABC-123)

## 🔄 Type of Change

- [ ] Feature
- [ ] Bugfix
- [ ] Refactor
- [ ] Hotfix
- [ ] Chore / Maintenance

## 🧪 How to Test

1. Call endpoint: `GET /api/...`
2. Use test data: `...`
3. Expected result:
   - 
    -

## ✅ Checklist

- [ ] Performed **self-review** of the code
- [ ] Updated **validations** if business logic changed
- [ ] Updated **API docs / Postman collection** if the endpoint changed
- [ ] All **tests are passing** (if applicable)
- [ ] No **breaking changes** for existing clients

## 🧱 Migration / Database Changes

- [ ] No database changes
- [ ] There is a new migration (describe below):

### CODE REVIEW

File: app/helpers.php [hunk 1]
```diff
@@ -427,11 +427,26 @@ function getTimeZoneByRequest()
 }
 function getTimeZoneBranchByRequest()
 {
-    $user = User::where('id', auth()->user()->id)
-   ->with(['userProfessionalData','userProfessionalData.branch','userProfessionalData.branch.address','userProfessionalData.branch.address.country','userProfessionalData.branch.address.country.timezones'])->first() ;
-    $timezone = $user?->userProfessionalData?->branch?->address?->country?->timezones ;
+    return once(function () {
```
Comment (AmrSaleh2234) [inline app/helpers.php:430]: The `once()` helper is being used here but there's no explanation of why caching is needed for this function. Consider adding documentation about the caching strategy if this is intentional.

File: app/helpers.php [hunk 2]
```diff
@@ -427,11 +427,26 @@ function getTimeZoneByRequest()
...
+        if (! auth()->check()) {
+            return config('app.timezone');
```
Comment (AmrSaleh2234) [inline app/helpers.php:432]: The fallback timezone `Asia/Riyadh` is hardcoded. Consider making this configurable via the application's configuration file.

File: app/helpers.php [hunk 3]
```diff
@@ -427,11 +427,26 @@ function getTimeZoneByRequest()
...
+            ->with([
+                'userProfessionalData',
+                'userProfessionalData.branch',
+                'userProfessionalData.branch.address',
+                'userProfessionalData.branch.address.country',
+                'userProfessionalData.branch.address.country.timezones',
+            ])
+            ->first();
+
+        $timezone = $user?->userProfessionalData?->branch?->address?->country?->timezones;
```
Comment (AmrSaleh2234) [inline app/helpers.php:446]: Accessing `timezones[0]` directly without checking if the array exists could lead to undefined index errors. Consider using `Arr::first($timezones)` or adding a null check.

File: modules/Attendance/Controllers/UserAttendanceController.php [hunk 4]
```diff
@@ -42,13 +43,11 @@ public function __construct(
...
-
-            $timezone = getTimeZoneBranchByRequest() ?? config('app.timezone');
-            $targetDate = $date ?? \Carbon\Carbon::now($timezone)->format('Y-m-d');
-
-            $result = $this->userAttendanceService->getUserConstraints($userId, $targetDate);
+            $user = $request->user();
+            if (!$user instanceof User) {
+                return Json::error('Unauthorized.', 401);
+            }
+            $result = $this->userAttendanceService->getUserConstraints($user, $request->input('date'));
```
Comment (AmrSaleh2234) [inline modules/Attendance/Controllers/UserAttendanceController.php:50]: The method call `getUserConstraints` now passes a User object instead of user ID and date string. Ensure the `UserAttendanceService::getUserConstraints` method has been updated to handle this change. Otherwise, this will cause runtime errors.

File: modules/Attendance/Controllers/UserAttendanceController.php [hunk 5]
```diff
@@ -42,13 +43,11 @@ public function __construct(
...
```
Comment (AmrSaleh2234) [inline modules/Attendance/Controllers/UserAttendanceController.php:51]: The timezone handling has been removed. This could lead to timezone-related issues when processing dates. Either add timezone handling back or ensure the service method handles timezones appropriately.

File: modules/Attendance/Controllers/UserAttendanceController.php [hunk 6]
```diff
@@ -42,13 +43,11 @@ public function __construct(
...
     {
         try {
-            $userId = (string) Auth::id();
-            $date = $request->input('date'); // Optional: Y-m-d format, defaults to today if null
-
-            $timezone = getTimeZoneBranchByRequest() ?? config('app.timezone');
-            $targetDate = $date ?? \Carbon\Carbon::now($timezone)->format('Y-m-d');
-
-            $result = $this->userAttendanceService->getUserConstraints($userId, $targetDate);
+            $user = $request->user();
```
Comment (AmrSaleh2234) [inline modules/Attendance/Controllers/UserAttendanceController.php:46]: The error message 'Unauthorized' is too generic. Consider providing a more specific message like 'Invalid user authentication'.

File: modules/Attendance/Controllers/UserAttendanceController.php (hunk 4 above)
Comment (AmrSaleh2234) [inline modules/Attendance/Controllers/UserAttendanceController.php:50]: The `$request->input('date')` is now passed directly without any default value or validation. This could lead to null values being passed to the service method. Either add a default value or ensure the service handles null dates properly.

File: modules/Attendance/Controllers/UserAttendanceController.php [hunk 7]
```diff
@@ -42,13 +43,11 @@ public function __construct(
...
+            if (!$user instanceof User) {
```
Comment (AmrSaleh2234) [inline modules/Attendance/Controllers/UserAttendanceController.php:47]: The type check for User instance should be moved to a middleware or request validation layer rather than handling it in the controller method.

File: modules/Attendance/Controllers/UserAttendanceController.php [hunk 8]
```diff
@@ -42,13 +43,11 @@ public function __construct(
...
+                return Json::error('Unauthorized.', 401);
```
Comment (AmrSaleh2234) [inline modules/Attendance/Controllers/UserAttendanceController.php:48]: The HTTP status code 401 should be extracted as a constant for better maintainability.

File: modules/Attendance/Controllers/UserAttendanceController.php (hunk 4 above)
Comment (AmrSaleh2234) [inline modules/Attendance/Controllers/UserAttendanceController.php:50]: Consider adding input validation for the date format before passing it to the service method.

Comment (AmrSaleh2234) [inline modules/Attendance/Controllers/UserAttendanceController.php:50]: The service method signature change (accepting User instead of string ID) should be clearly documented in the service interface to avoid confusion.

Comment (AmrSaleh2234) [inline modules/Attendance/Controllers/UserAttendanceController.php:50]: If the service method still expects a formatted date string, you should format the date before passing it to avoid potential type mismatches.

Comment (AmrSaleh2234) [inline modules/Attendance/Controllers/UserAttendanceController.php:50]: Consider adding a test case for when the date parameter is null to ensure proper handling.

Comment (AmrSaleh2234) [inline modules/Attendance/Controllers/UserAttendanceController.php:50]: The removal of timezone handling should be documented as it might affect clients relying on timezone-specific behavior.

Comment (AmrSaleh2234) [inline modules/Attendance/Controllers/UserAttendanceController.php:50]: Ensure all calling code has been updated to match the new parameter requirements of `getUserConstraints`.

Comment (AmrSaleh2234) [inline modules/Attendance/Controllers/UserAttendanceController.php:50]: The change from string ID to User object should be reflected in any API documentation.

Comment (AmrSaleh2234) [inline modules/Attendance/Controllers/UserAttendanceController.php:50]: Consider whether the User object might contain sensitive data that shouldn't be passed around in service methods.

Comment (AmrSaleh2234) [inline modules/Attendance/Controllers/UserAttendanceController.php:50]: If the service method performs any caching based on user ID, the cache keys might need updating to handle User objects instead of string IDs.

Comment (AmrSaleh2234) [inline modules/Attendance/Controllers/UserAttendanceController.php:50]: The removal of explicit timezone handling might affect reporting consistency across timezones. Verify this change with stakeholders.

Comment (AmrSaleh2234) [inline modules/Attendance/Controllers/UserAttendanceController.php:50]: Consider adding a deprecation notice if there are other parts of the system still using the old string ID approach.

File: modules/Attendance/Repositories/AttendanceConstraintRepository.php [hunk 9]
```diff
@@ -79,6 +86,9 @@ public function updateConstraint(UuidInterface $id, array $data): AttendanceCons
      */
     public function deleteConstraint(UuidInterface $id): bool
     {
+        $constraint = $this->getConstraint($id);
+        $this->assertNoOpenAttendanceForConstraint($id->toString(), (string) $constraint->company_id);
+
         return $this->delete($id);
```
Comment (AmrSaleh2234) [inline modules/Attendance/Repositories/AttendanceConstraintRepository.php:92]: Should wrap the delete operation in a try-catch block to handle potential database exceptions gracefully.

File: modules/Attendance/Repositories/AttendanceConstraintRepository.php [hunk 10]
```diff
@@ -213,6 +223,55 @@ private function getPaginationData($query, int $page, int $perPage): array
...
+    private function assertNoOpenAttendanceForConstraint(string $constraintId, string $companyId): void
+    {
+        if ($this->hasOpenAttendanceForConstraint($constraintId, $companyId)) {
+            throw AttendanceException::cannotModifyConstraintWithOpenAttendance();
+        }
+    }
+
+    private function hasOpenAttendanceForConstraint(string $constraintId, string $companyId): bool
+    {
+        $openAttendance = static function ($query) {
```
Comment (AmrSaleh2234) [inline modules/Attendance/Repositories/AttendanceConstraintRepository.php:238]: The closure `$openAttendance` is defined inside the method but used multiple times. Consider extracting it to a class property or method to avoid redefinition.

File: modules/Attendance/Repositories/AttendanceConstraintRepository.php [hunk 11]
```diff
@@ -213,6 +223,55 @@ private function getPaginationData($query, int $page, int $perPage): array
...
+            $query->whereNotNull('clock_in_time')->whereNull('clock_out_time');
+        };
+
+        if (AppliedAttendanceConstraint::query()
```
Comment (AmrSaleh2234) [inline modules/Attendance/Repositories/AttendanceConstraintRepository.php:242]: The query chain could be simplified by combining the conditions:
```php
AppliedAttendanceConstraint::query()
    ->where('company_id', $companyId)
    ->where('constraint_snapshot->id', $constraintId)
    ->whereHas('attendance', $openAttendance)
    ->exists()
```

File: modules/Attendance/Repositories/AttendanceConstraintRepository.php [hunk 12]
```diff
@@ -213,6 +223,55 @@ private function getPaginationData($query, int $page, int $perPage): array
...
+            ->where('constraint_snapshot->id', $constraintId)
+            ->exists()) {
+            return true;
+        }
+
+        $pivotUserIds = DB::table('attendance_constraint_user')
+            ->where('attendance_constraint_id', $constraintId)
+            ->pluck('user_id');
+
+        if ($pivotUserIds->isNotEmpty()
```
Comment (AmrSaleh2234) [inline modules/Attendance/Repositories/AttendanceConstraintRepository.php:254]: The condition `$pivotUserIds->isNotEmpty()` could be merged with the subsequent query using `when()`:
```php
Attendance::query()
    ->where('company_id', $companyId)
    ->where($openAttendance)
    ->when($pivotUserIds->isNotEmpty(), fn($q) => $q->whereIn('user_id', $pivotUserIds))
    ->exists()
```

File: modules/Attendance/Repositories/AttendanceConstraintRepository.php [hunk 13]
```diff
@@ -213,6 +223,55 @@ private function getPaginationData($query, int $page, int $perPage): array
...
+                ->whereIn('user_id', $pivotUserIds)
+                ->exists()) {
+            return true;
+        }
+
+        $professionalUserIds = UserProfessionalData::query()
+            ->where('attendance_constraint_id', $constraintId)
+            ->pluck('user_id');
+
+        return $professionalUserIds->isNotEmpty()
```
Comment (AmrSaleh2234) [inline modules/Attendance/Repositories/AttendanceConstraintRepository.php:267]: Similar to previous comment, the condition could be merged using `when()`:
```php
Attendance::query()
    ->where('company_id', $companyId)
    ->where($openAttendance)
    ->when($professionalUserIds->isNotEmpty(), fn($q) => $q->whereIn('user_id', $professionalUserIds))
    ->exists()
```

File: modules/Attendance/Services/AttendanceConstraintService.php [hunk 14]
```diff
@@ -65,7 +65,14 @@ public function __construct(
     public function validateAttendance(Attendance $attendance, array $requestData = [],bool $isDryRun = false): array
     {
         $violations = [];
-        $user = $attendance->user;
+        // Get user ID from attendance to avoid readonly property issue
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/AttendanceConstraintService.php:68]: The comment `// Get user ID from attendance to avoid readonly property issue` is unnecessary as the code change is self-explanatory. Consider removing it.

File: modules/Attendance/Services/AttendanceConstraintService.php [hunk 15]
```diff
@@ -65,7 +65,14 @@ public function __construct(
...
+        $userId = $attendance->user_id;
+        $user = User::find($userId);
+        
+        if (!$user) {
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/AttendanceConstraintService.php:72]: Returning an empty array when user is not found might silently hide an issue. Consider throwing an appropriate exception instead.

File: modules/Attendance/Services/AttendanceConstraintService.php [hunk 16]
```diff
@@ -115,47 +122,8 @@ public function validateAttendance(Attendance $attendance, array $requestData =
      */
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/AttendanceConstraintService.php:122]: The method `getEffectiveConstraintForUser` is being simplified to call `getApplicableConstraintsForDataRetrieval`, but there's no explanation why this change is being made. The logic appears to be significantly different from the original implementation which had specific priority rules (default constraints first, then others). This could potentially change business logic behavior.

File: modules/Attendance/Services/AttendanceConstraintService.php [hunk 17]
```diff
@@ -115,47 +122,8 @@ public function validateAttendance(Attendance $attendance, array $requestData =
...
-
-                // Condition C: Global constraints (not assigned to any specific user or branch).
-                ->orWhere(function ($q) {
-                    $q->where(fn($sub) => $sub->whereNull('user_ids')->orWhereJsonLength('user_ids', 0))
-                      ->where(fn($sub) => $sub->whereNull('branch_ids')->orWhereJsonLength('branch_ids', 0));
-                });
-            })
-            ->get();
+        // Use the same optimized logic as getApplicableConstraintsForDataRetrieval
+        return $this->getApplicableConstraintsForDataRetrieval($user);
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/AttendanceConstraintService.php:126]: The method `getApplicableConstraintsForDataRetrieval` is being called but its implementation isn't shown in the diff. This makes it impossible to verify if the new simplified approach maintains the same business logic requirements as the original implementation.

File: modules/Attendance/Services/AttendanceConstraintService.php [hunk 18]
```diff
@@ -474,8 +442,14 @@ public function validateBreakEnd(Attendance $attendance): bool|array
             return false;
         }
 
+        // Get user to avoid readonly property issue
+        $user = User::find($attendance->user_id);
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/AttendanceConstraintService.php:446]: Avoid using `User::find()` directly in the service layer. Consider injecting UserRepository or similar dependency through constructor instead.

File: modules/Attendance/Services/AttendanceConstraintService.php [hunk 19]
```diff
@@ -474,8 +442,14 @@ public function validateBreakEnd(Attendance $attendance): bool|array
...
+        if (!$user) {
+            return false;
+        }
+        
         // Get all applicable constraints for the user
-        $constraints = $this->getApplicableConstraints($attendance->user);
+        $constraints = $this->getApplicableConstraints($user);
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/AttendanceConstraintService.php:452]: The line number in the diff (478) doesn't match the actual line number (452). Please verify the diff is correctly showing the changes.

File: modules/Attendance/Services/AttendanceConstraintService.php [hunk 20]
```diff
@@ -915,14 +890,15 @@ public function getApplicableConstraintsForDataRetrieval(User $user): Collection
         $userBranchId = $user->userProfessionalData?->branch?->id;
         $userDepartmentId = $user->userProfessionalData?->department?->id;
 
-        // 1. Check for a Default Constraint on the Branch
+        // 1. Check for a Default Constraint on the Branch (single query)
         if ($userBranchId) {
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/AttendanceConstraintService.php:894]: The comment `(single query)` is redundant since the code clearly shows a single query. Remove this comment.

File: modules/Attendance/Services/AttendanceConstraintService.php [hunk 21]
```diff
@@ -915,14 +890,15 @@ public function getApplicableConstraintsForDataRetrieval(User $user): Collection
...
             $defaultConstraint = AttendanceConstraint::whereHas('branches', function ($query) use ($userBranchId) {
                 $query->where('management_hierarchies.id', $userBranchId)
-                      ->where('is_default', true); // Assuming a pivot or relationship for default
+                      ->where('is_default', true);
             })
+            ->where('company_id', $user->company_id)
             ->where('is_active', true)
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/AttendanceConstraintService.php:900]: The newly added `where('company_id', $user->company_id)` condition should be documented in the PR description since it changes the business logic by adding an additional filter.

File: modules/Attendance/Services/AttendanceService.php [hunk 22]
```diff
@@ -100,9 +101,61 @@ public function clockIn(ClockInDTO $clockInDTO)
...
+        foreach ($constraints['all_work_periods'] ?? [] as $period) {
+            $pStart = $period['period_start_time_carbon'] ?? null;
+            if (!$pStart instanceof Carbon) {
+                $date = $period['date'] ?? $currentPeriodEnd->format('Y-m-d');
+                $startTime = $period['start_time'] ?? null;
+                if (!is_string($startTime) || $startTime === '') {
+                    continue;
+                }
+                $timeHi = strlen($startTime) > 5 ? substr($startTime, 0, 5) : $startTime;
+                $pStart = Carbon::createFromFormat('Y-m-d H:i', $date . ' ' . $timeHi, $tz);
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/AttendanceService.php:146]: Consider validating the `$timeHi` format before using it in `Carbon::createFromFormat` to prevent potential DateTime exceptions.

File: modules/Attendance/Services/AttendanceService.php [hunk 23]
```diff
@@ -100,9 +101,61 @@ public function clockIn(ClockInDTO $clockInDTO)
...
+     * Queue auto clock-out at the start of the next scheduled period (same merged day list from constraints)
+     * if the user is still clocked into this shift.
+     */
+    private function scheduleAutoClockOutWhenNextShiftStarts(
+        Attendance $attendance,
+        array $constraints,
+        Carbon $currentPeriodEnd
+    ): void {
+        $nextStart = $this->resolveNextShiftStartAfterPeriodEnd($constraints, $currentPeriodEnd);
+        if ($nextStart === null || !$nextStart->isFuture()) {
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/AttendanceService.php:119]: The condition `!$nextStart->isFuture()` seems redundant since `$nextStart` is derived from periods that should always be in the future (greater than `$currentPeriodEnd`). Consider removing this check.

File: modules/Attendance/Services/AttendanceService.php [hunk 24]
```diff
@@ -100,9 +101,61 @@ public function clockIn(ClockInDTO $clockInDTO)
...
+    }
+
+    /**
+     * @param  array<string, mixed>  $constraints  Output of {@see AttendanceConstraintService::getTodaysWorkRulesForUser}
+     */
+    private function resolveNextShiftStartAfterPeriodEnd(array $constraints, Carbon $currentPeriodEnd): ?Carbon
+    {
+        $tz = $currentPeriodEnd->getTimezone();
+
+        foreach ($constraints['all_work_periods'] ?? [] as $period) {
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/AttendanceService.php:137]: Add a null check for `$constraints['all_work_periods']` before the foreach loop to prevent potential undefined index warnings.

File: modules/Attendance/Services/AttendanceService.php [hunk 25]
```diff
@@ -233,9 +286,9 @@ public function endBreak(UuidInterface|string $userId, ?string $notes = null): A
     /**
      * Get current attendance for user
      */
-    public function getCurrentAttendance(UuidInterface $userId): ?Attendance
+    public function getCurrentAttendance(UuidInterface $userId, bool $withUser = true): ?Attendance
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/AttendanceService.php:289]: The parameter `$withUser` should be documented with a `@param` tag in the method docblock to explain its purpose.

File: modules/Attendance/Services/AttendanceService.php [hunk 26]
```diff
@@ -233,9 +286,9 @@ public function endBreak(UuidInterface|string $userId, ?string $notes = null): A
...
     {
-        return $this->attendanceRepository->getCurrentAttendance($userId);
+        return $this->attendanceRepository->getCurrentAttendance($userId, $withUser);
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/AttendanceService.php:291]: The method call should maintain consistency with the parameter name - either rename the parameter to match the repository method (`$withUser`) or vice versa.

File: modules/Attendance/Services/UserAttendanceService.php [hunk 27]
```diff
@@ -88,29 +149,37 @@ public function checkClockInStatus(UuidInterface|string $userId): array
...
+     * One query: attendances for the target day (by start_time / clock_in_time) plus any still-open shift
+     * (clock_in set, clock_out null) so overnight sessions are included without a second DB round-trip.
      *
-     * @param User $user
-     * @param Carbon $date
-     * @return Collection
+     * @return array{0: Collection<int, Attendance>, 1: Attendance|null}
      */
-    private function getAttendancesForDate(User $user, Carbon $date): Collection
+    private function fetchDayAttendancesAndCurrentOpen(User $user, Carbon $date): array
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/UserAttendanceService.php:157]: Method name `fetchDayAttendancesAndCurrentOpen` is unclear about what 'current open' refers to. Consider renaming to something more descriptive like `fetchDayAttendancesWithOpenShifts`.

File: modules/Attendance/Services/UserAttendanceService.php [hunk 28]
```diff
@@ -88,29 +149,37 @@ public function checkClockInStatus(UuidInterface|string $userId): array
...
+                    $inner->whereBetween('start_time', [$dayStartUtc, $dayEndUtc])
+                        ->orWhereBetween('clock_in_time', [$dayStartUtc, $dayEndUtc]);
+                })->orWhere(function ($inner) {
+                    $inner->whereNotNull('clock_in_time')->whereNull('clock_out_time');
+                });
             })
             ->orderBy('start_time')
             ->get();
+
+        $currentOpen = $records->first(static function (Attendance $a) {
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/UserAttendanceService.php:178]: The logic for finding `$currentOpen` could be simplified using `Collection::firstWhere()`:
```php
$currentOpen = $records->firstWhere('clock_out_time', null);
```

File: modules/Attendance/Services/UserAttendanceService.php [hunk 29]
```diff
@@ -88,29 +149,37 @@ public function checkClockInStatus(UuidInterface|string $userId): array
...
+            return $a->clock_in_time !== null && $a->clock_out_time === null;
+        });
+
+        return [$records, $currentOpen];
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/UserAttendanceService.php:182]: Returning an array with positional elements (`[0]` and `[1]`) makes the code harder to understand. Consider returning an associative array with named keys or using a DTO.

File: modules/Attendance/Services/UserAttendanceService.php [hunk 30]
```diff
@@ -121,24 +190,103 @@ private function getAttendancesForDate(User $user, Carbon $date): Collection
...
-            $periodStart = $this->parsePeriodTime($period, 'start', $date);
-            $periodEnd = $this->parsePeriodTime($period, 'end', $date);
+    private function enhancePeriodsWithAttendance(
+        array $periods,
+        Collection $attendances,
+        Carbon $date,
+        array $earlyClockInRules,
+        ?Attendance $currentAttendance = null
+    ): array {
+        $timezone = $this->getTimezone();
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/UserAttendanceService.php:200]: Consider caching the timezone retrieval if this method is called frequently, as `getTimezone()` might involve some overhead.

File: modules/Attendance/Services/UserAttendanceService.php [hunk 31]
```diff
@@ -121,24 +190,103 @@ private function getAttendancesForDate(User $user, Carbon $date): Collection
...
+        $now = Carbon::now($timezone);
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/UserAttendanceService.php:201]: The `now` timestamp should be injected as a parameter to make the method more testable and avoid relying on the system clock.

File: modules/Attendance/Services/UserAttendanceService.php [hunk 32]
```diff
@@ -121,24 +190,103 @@ private function getAttendancesForDate(User $user, Carbon $date): Collection
...
+
+        $periodBounds = [];
+        foreach ($periods as $idx => $period) {
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/UserAttendanceService.php:204]: The `foreach` loop could potentially be replaced with `array_map` for better readability and functional style.

File: modules/Attendance/Services/UserAttendanceService.php [hunk 33]
```diff
@@ -121,24 +190,103 @@ private function getAttendancesForDate(User $user, Carbon $date): Collection
...
+            $periodBounds[$idx] = [
+                'start' => $this->parsePeriodTime($period, 'start', $date),
+                'end' => $this->parsePeriodTime($period, 'end', $date),
+            ];
+        }
+
+        $activePeriodIndex = $this->resolveSingleActivePeriodIndex(
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/UserAttendanceService.php:211]: The method name `resolveSingleActivePeriodIndex` is quite long. Consider shortening it while keeping it descriptive.

File: modules/Attendance/Services/UserAttendanceService.php [hunk 34]
```diff
@@ -121,24 +190,103 @@ private function getAttendancesForDate(User $user, Carbon $date): Collection
...
+     *
+     * @param array<int, array{start: Carbon, end: Carbon}> $periodBounds
+     */
+    private function resolveSingleActivePeriodIndex(
+        array $periodBounds,
+        Carbon $now,
+        array $earlyClockInRules,
+        ?Attendance $currentAttendance
+    ): ?int {
+        if ($currentAttendance !== null) {
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/UserAttendanceService.php:256]: The null check for `$currentAttendance` should be moved to the beginning of the method to fail fast.

File: modules/Attendance/Services/UserAttendanceService.php [hunk 35]
```diff
@@ -121,24 +190,103 @@ private function getAttendancesForDate(User $user, Carbon $date): Collection
...
+        Attendance $attendance,
+        Carbon $periodStart,
+        Carbon $periodEnd
+    ): bool {
+        if (!$attendance->clock_in_time) {
+            return false;
+        }
+
+        $attendanceTz = $attendance->timezone ?? $periodStart->getTimezone();
+        $clockInCarbon = $attendance->clock_in_time instanceof Carbon
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/UserAttendanceService.php:283]: The timezone handling logic is complex. Consider extracting it into a separate method for better maintainability.

File: modules/Attendance/Services/UserAttendanceService.php [hunk 36]
```diff
@@ -121,24 +190,103 @@ private function getAttendancesForDate(User $user, Carbon $date): Collection
...
+            ? $attendance->clock_in_time->copy()->setTimezone($attendanceTz)
+            : Carbon::parse($attendance->clock_in_time, $attendanceTz);
+
+        $clockInInPeriodTz = $clockInCarbon->copy()->setTimezone($periodStart->getTimezone());
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/UserAttendanceService.php:287]: The timezone conversion is happening twice (lines 283 and 287). This could be optimized.

File: modules/Attendance/Services/UserAttendanceService.php [hunk 37]
```diff
@@ -263,11 +397,19 @@ private function formatAttendanceForPeriod(Attendance $attendance, Carbon $perio
      * @param array $period
      * @param float $totalWorkHours
      * @param array $attendance
-     * @param bool $isActive
+     * @param bool $isActiveByTime Now inside this period or its early clock-in window (drives can_clock_in)
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/UserAttendanceService.php:400]: The parameter name `isActiveByTime` is inconsistent with the method's parameter naming convention (camelCase vs snake_case). Consider renaming to `isActiveByTime` to match the existing style.

File: modules/Attendance/Services/UserAttendanceService.php [hunk 38]
```diff
@@ -263,11 +397,19 @@ private function formatAttendanceForPeriod(Attendance $attendance, Carbon $perio
...
+     * @param bool $isActiveForDisplay Single "current" period for UI (open shift period, else time-based)
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/UserAttendanceService.php:401]: The parameter name `isActiveForDisplay` is inconsistent with the method's parameter naming convention (camelCase vs snake_case). Consider renaming to `isActiveForDisplay` to match the existing style.

File: modules/Attendance/Services/UserAttendanceService.php [hunk 39]
```diff
@@ -263,11 +397,19 @@ private function formatAttendanceForPeriod(Attendance $attendance, Carbon $perio
...
-    private function mergePeriodData(array $period, float $totalWorkHours, array $attendance, bool $isActive, array $earlyClockInRules): array
-    {
+    private function mergePeriodData(
+        array $period,
+        float $totalWorkHours,
+        array $attendance,
+        bool $isActiveByTime,
+        bool $isActiveForDisplay,
+        array $earlyClockInRules,
+        ?Attendance $currentAttendance = null
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/UserAttendanceService.php:411]: The new parameter `currentAttendance` is nullable but there's no null check in the method body. Either add null handling or make it required if it's essential for the method's operation.

File: modules/Attendance/Services/UserAttendanceService.php [hunk 40]
```diff
@@ -279,19 +421,15 @@ private function mergePeriodData(array $period, float $totalWorkHours, array $at
         $hasActiveAttendance = collect($attendance)->contains(function ($att) {
             return $att['status'] === 'active';
         });
-        
-        $getCurrentAttendance = $this->attendanceService->getCurrentAttendance(auth()->user()->id);
-        $canClockIn = $isActive && !$hasActiveAttendance && (bool) !$getCurrentAttendance;
-
 
-        $effectiveIsActive = $isActive || $hasActiveAttendance;
+        $canClockIn = $isActiveByTime && ! $hasActiveAttendance && $currentAttendance === null;
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/UserAttendanceService.php:425]: The condition `$isActiveByTime && ! $hasActiveAttendance && $currentAttendance === null` should be consistent with the previous logic. Consider renaming `$isActiveByTime` to `$isActive` for clarity since it's used in similar context before.

File: modules/Attendance/Services/UserAttendanceService.php [hunk 41]
```diff
@@ -279,19 +421,15 @@ private function mergePeriodData(array $period, float $totalWorkHours, array $at
...
 
         return array_merge($cleanedPeriod, [
             'total_work_hours' => $totalWorkHours,
-            'is_active' => $effectiveIsActive,
+            'is_active' => $isActiveForDisplay,
             'total_hours_present' => round($totalHoursPresent, 2),
             'can_clock_in' => $canClockIn,
-            'can_clock_out' => (bool) $getCurrentAttendance,
+            'can_clock_out' => $currentAttendance !== null && $isActiveForDisplay,
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/UserAttendanceService.php:432]: The condition `$currentAttendance !== null && $isActiveForDisplay` introduces a new variable `$isActiveForDisplay` which wasn't in the original logic. Ensure this variable is properly defined and its usage aligns with the business requirements.

File: modules/Attendance/Services/UserAttendanceService.php [hunk 42]
```diff
@@ -279,19 +421,15 @@ private function mergePeriodData(array $period, float $totalWorkHours, array $at
...
-        $canClockIn = $isActive && !$hasActiveAttendance && (bool) !$getCurrentAttendance;
-
 
-        $effectiveIsActive = $isActive || $hasActiveAttendance;
+        $canClockIn = $isActiveByTime && ! $hasActiveAttendance && $currentAttendance === null;
 
         return array_merge($cleanedPeriod, [
             'total_work_hours' => $totalWorkHours,
-            'is_active' => $effectiveIsActive,
+            'is_active' => $isActiveForDisplay,
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/UserAttendanceService.php:429]: The variable `$isActiveForDisplay` replaces `$effectiveIsActive` but their logic seems different (`$isActive || $hasActiveAttendance` vs new variable). Verify this change doesn't introduce unintended behavior.

File: modules/Attendance/Services/UserAttendanceService.php [hunk 43]
```diff
@@ -396,7 +534,7 @@ private function getCurrentAttendanceSafely(UuidInterface|string $userId): ?Atte
     {
         try {
             $userIdUuid = is_string($userId) ? Uuid::fromString($userId) : $userId;
-            return $this->attendanceService->getCurrentAttendance($userIdUuid);
+            return $this->attendanceService->getCurrentAttendance($userIdUuid, false);
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/UserAttendanceService.php:537]: The second parameter `false` was added without explanation. Since this changes the behavior of the method, we should:
1. Document this change in the PR description
2. Ensure this aligns with the expected behavior in all calling contexts
3. Consider if this should be configurable via method parameter rather than hardcoded

File: modules/Attendance/Services/UserAttendanceService.php [hunk 44]
```diff
@@ -639,6 +777,23 @@ private function formatMinutesToTime(int $minutes): string
...
+     * Prefer timezone from the user's branch (already eager-loaded) to avoid a second User query via getTimeZoneBranchByRequest().
+     */
+    private function timezoneFromUserBranch(User $user): string
+    {
+        $timezones = $user->userProfessionalData?->branch?->address?->country?->timezones;
+        if (is_array($timezones) && isset($timezones[0]['zoneName']) && is_string($timezones[0]['zoneName'])) {
+            return $timezones[0]['zoneName'];
+        }
+
         return getTimeZoneBranchByRequest() ?? config('app.timezone');
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/UserAttendanceService.php:797]: The fallback logic (`getTimeZoneBranchByRequest() ?? config('app.timezone')`) is duplicated from the `getTimezone()` method above. Consider extracting this into a private method to avoid code duplication.

File: modules/Attendance/Services/UserAttendanceService.php [hunk 45]
```diff
@@ -639,6 +777,23 @@ private function formatMinutesToTime(int $minutes): string
...
+        return getTimeZoneBranchByRequest() ?? config('app.timezone');
+    }
+
+    /**
+     * Prefer timezone from the user's branch (already eager-loaded) to avoid a second User query via getTimeZoneBranchByRequest().
+     */
+    private function timezoneFromUserBranch(User $user): string
+    {
+        $timezones = $user->userProfessionalData?->branch?->address?->country?->timezones;
+        if (is_array($timezones) && isset($timezones[0]['zoneName']) && is_string($timezones[0]['zoneName'])) {
```
Comment (AmrSaleh2234) [inline modules/Attendance/Services/UserAttendanceService.php:793]: The nested null checks (`userProfessionalData?->branch?->address?->country?->timezones`) suggest a potential code smell. Consider:
1. Adding validation that these relationships exist when the User is created/updated
2. Or moving this timezone resolution logic to a dedicated service class
